runners/spark/src/main/java/org/apache/beam/runners/spark/util/SparkCompat.java

```diff
@@ -0,0 +1,241 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.beam.runners.spark.util;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Iterator;
+import java.util.Map;
+import org.apache.spark.SparkEnv;
+import org.apache.spark.storage.BlockId;
+import org.apache.spark.storage.BlockResult;
+import org.apache.spark.storage.StorageLevel;
+import scala.Option;
+import scala.Tuple2;
+import scala.reflect.ClassTag;
+
+/** A class with helper functions and utilities for compatibility between spark versions. */
+public final class SparkCompat {
+
+  private SparkCompat() {}
+
+  /**
+   * A class that wraps an {@link Iterator} and exposes it as an {@link Iterable} or wraps an {@link
+   * Iterable} and exposes it as an {@link Iterator}. In the second case it creates an Iterator by
+   * calling the iterable.Iterator method and then respects its semantics.
+   *
+   * <p>Notice that both uses cannot be combined, so it is up to the user to respect the logic since
+   * we cannot ensure this.
+   *
+   * <p>This is an internal class used to ensure compatibility between spark versions and should not
+   * be used for other purposes.
+   */
+  public static class IterableIterator<T> implements Iterable<T>, Iterator<T> {
+
+    private final Iterable<T> delegate;
+    private final Iterator<T> iter;
+
+    public IterableIterator(Iterable<T> delegate) {
```

Comment:
Cool.

In favor of consistency with Guava perhaps the ctors could be turned into static `of(...)` methods so that it reads `IterableIterator.of(itr)`.